src/object/cli_obj.c

```diff
@@ -3728,9 +3728,17 @@ obj_shard_comp_cb(tse_task_t *task, struct shard_auxi_args *shard_auxi,
 				shard_auxi->shard);
 		}
 	} else if (obj_retry_error(ret)) {
+		int rc;
+
 		D_DEBUG(DB_IO, "shard %d ret %d.\n", shard_auxi->shard, ret);
 		if (obj_auxi->result == 0)
 			obj_auxi->result = ret;
+		rc = obj_auxi_add_failed_tgt(obj_auxi, shard_auxi->target);
```

Comment:
for temporary timeout failure, may need not to add it to failed tgt? 

BTW, for the issue this PR try to resolve "Failed targets should be kept and checked during retry, otherwise same target might be counted duplicately, and cause data lost."
In code obj_ec_fail_info_insert() we already with code to check duplicated tgt -
1384 >       if (obj_ec_tgt_in_err(fail_info->efi_tgt_list, fail_info->efi_ntgts, fail_tgt))             
1385 >       >       return 0;  

So even if we insert same tgt multiple times, it won't be counted more than one time. no?

Reply:
> So even if we insert same tgt multiple times, it won't be counted more than one time. no?

NVME IO retry will retry all shards initially, so if we do not remember the failure tgts, then innocent shards might be added to the failure list, then IO might be sent to the "wrong" server again.